File: bindings/python/ainos/stream.py
"""
Ainos SDK - Streaming Inference
================================

Provides the streaming iterator for consuming token-by-token inference
responses from the Ainos daemon.

The Ainos daemon supports Server-Sent Events (SSE)-style streaming over the
NDJSON protocol. When a request is sent with ``stream=True``, the daemon
responds with multiple ``"stream"`` type messages, each containing a single
token or chunk of the generated output.

Usage::

    async for chunk in client.infer_stream("my-model", "Hello world"):
        print(chunk.token, end="", flush=True)
        if chunk.final:
            print(f"\\nDone. Reason: {chunk.finish_reason}")

Architecture::

    ┌──────────────┐     ┌─────────────────┐     ┌──────────────┐
    │  AinosClient │────▶│  StreamIterator  │────▶│  asyncio.Queue│
    │              │     │                  │     │  (chunks)    │
    └──────────────┘     └─────────────────┘     └──────────────┘
"""

from __future__ import annotations

import asyncio
import logging
import typing as t
from dataclasses import dataclass, field

from ainos.errors import (
    InferenceError,
    StreamError,
    StreamInterruptedError,
    StreamNotStartedError,
)
from ainos.types import (
    InferenceChunk,
    JSONObject,
    RequestMessage,
    TokenUsage,
)
from ainos.utils import generate_request_id, json_encode, timestamp

log: logging.Logger = logging.getLogger("ainos.stream")

#: Maximum number of chunks that can be buffered in the stream queue.
_MAX_QUEUE_SIZE: int = 4096


# ---------------------------------------------------------------------------
# Stream events
# ---------------------------------------------------------------------------


@dataclass
class StreamEvent:
    """An event that occurs during streaming.

    This is used internally by the stream iterator to communicate between
    the transport layer and the consumer.

    Attributes:
        type: The event type (``"chunk"``, ``"error"``, or ``"done"``).
        chunk: The inference chunk (for ``"chunk"`` events).
        error: The error (for ``"error"`` events).
        final_data: The final response data (for ``"done"`` events).
    """

    type: str  # "chunk", "error", "done"
    chunk: t.Optional[InferenceChunk] = None
    error: t.Optional[BaseException] = None
    final_data: t.Optional[JSONObject] = None


# ---------------------------------------------------------------------------
# StreamIterator
# ---------------------------------------------------------------------------


class StreamIterator:
    """Async iterator that yields inference chunks from a streaming response.

    This class is not instantiated directly; instead, it is returned by
    ``AinosClient.infer_stream()``.

    The iterator reads from an internal ``asyncio.Queue`` that is populated
    by the transport layer's message handler.

    Attributes:
        request_id: The UUID of the streaming request.
        model: The model identifier.
        stream_started: Whether the stream has been started.
        stream_finished: Whether the stream has completed.
        token_count: Number of tokens received so far.
    """

    def __init__(
        self,
        request_id: str,
        model: str,
        queue: asyncio.Queue[StreamEvent],
        *,
        timeout: float = 60.0,
    ) -> None:
        """Initialise the stream iterator.

        Args:
            request_id: The UUID of the streaming request.
            model: The model identifier.
            queue: The queue that receives stream events from the transport.
            timeout: Maximum time to wait for the next chunk.
        """
        self.request_id: str = request_id
        self.model: str = model
        self._queue: asyncio.Queue[StreamEvent] = queue
        self._timeout: float = timeout

        self._stream_started: bool = False
        self._stream_finished: bool = False
        self._token_count: int = 0
        self._start_time: float = 0.0
        self._total_time: float = 0.0
        self._generated_text: t.List[str] = []

    # ------------------------------------------------------------------
    # Properties
    # ------------------------------------------------------------------

    @property
    def stream_started(self) -> bool:
        """Whether the stream has started receiving chunks."""
        return self._stream_started

    @property
    def stream_finished(self) -> bool:
        """Whether the stream has completed."""
        return self._stream_finished

    @property
    def token_count(self) -> int:
        """Number of tokens received so far."""
        return self._token_count

    @property
    def elapsed(self) -> float:
        """Elapsed time since the stream started, in seconds.

        Returns:
            Elapsed time in seconds, or 0.0 if the stream has not started.
        """
        if self._start_time == 0.0:
            return 0.0
        if self._stream_finished:
            return self._total_time
        return timestamp() - self._start_time

    @property
    def generated_text(self) -> str:
        """The full text generated so far (concatenation of all tokens)."""
        return "".join(self._generated_text)

    # ------------------------------------------------------------------
    # Async iterator protocol
    # ------------------------------------------------------------------

    def __aiter__(self) -> "StreamIterator":
        """Return self as the async iterator."""
        if not self._stream_started:
            self._stream_started = True
            self._start_time = timestamp()
        return self

    async def __anext__(self) -> InferenceChunk:
        """Get the next chunk from the stream.

        Returns:
            The next InferenceChunk.

        Raises:
            StopAsyncIteration: When the stream has completed.
            StreamInterruptedError: If the stream is interrupted.
            StreamError: If a stream error occurs.
            asyncio.TimeoutError: If no chunk is received within the timeout.
        """
        if self._stream_finished:
            raise StopAsyncIteration

        try:
            event: StreamEvent = await asyncio.wait_for(
                self._queue.get(),
                timeout=self._timeout,
            )
        except asyncio.TimeoutError:
            self._stream_finished = True
            raise

        if event.type == "done":
            self._stream_finished = True
            self._total_time = timestamp() - self._start_time
            raise StopAsyncIteration

        if event.type == "error":
            self._stream_finished = True
            self._total_time = timestamp() - self._start_time
            if event.error is not None:
                raise event.error
            raise StreamError("Unknown stream error")

        if event.type == "chunk" and event.chunk is not None:
            chunk: InferenceChunk = event.chunk
            self._token_count += 1
            if chunk.token:
                self._generated_text.append(chunk.token)

            if chunk.final:
                self._stream_finished = True
                self._total_time = timestamp() - self._start_time

            return chunk

        raise StreamError(f"Unknown stream event type: {event.type}")

    # ------------------------------------------------------------------
    # Utility methods
    # ------------------------------------------------------------------

    async def collect(self) -> str:
        """Collect all tokens from the stream into a single string.

        This is a convenience method that iterates over the entire stream
        and returns the concatenated text.

        Returns:
            The full generated text.

        Example:
            text = await client.infer_stream("model", "Hello").collect()
        """
        result: t.List[str] = []
        async for chunk in self:
            result.append(chunk.token)
        return "".join(result)

    async def cancel(self) -> None:
        """Cancel the stream early.

        This stops the stream and prevents further chunks from being
        processed. The stream is marked as finished.
        """
        self._stream_finished = True
        self._total_time = timestamp() - self._start_time
        log.info("Stream %s cancelled after %d tokens", self.request_id, self._token_count)

    def get_stats(self) -> t.Dict[str, t.Any]:
        """Get stream statistics.

        Returns:
            A dictionary of stream statistics.
        """
        return {
            "request_id": self.request_id,
            "model": self.model,
            "token_count": self._token_count,
            "elapsed": self.elapsed,
            "stream_started": self._stream_started,
            "stream_finished": self._stream_finished,
            "generated_text_length": len(self.generated_text),
        }

    def __repr__(self) -> str:
        """Return a string representation of the stream iterator."""
        return (
            f"StreamIterator(request_id={self.request_id}, "
            f"model={self.model}, "
            f"tokens={self._token_count}, "
            f"finished={self._stream_finished})"
        )


# ---------------------------------------------------------------------------
# Stream manager
# ---------------------------------------------------------------------------


class StreamManager:
    """Manages multiple active streams.

    The StreamManager keeps track of all active streams, allowing the client
    to manage and monitor concurrent streaming requests.

    Usage::

        manager = StreamManager()
        iterator = manager.create_stream(request_id, model)
        # ... later ...
        active = manager.list_active_streams()
    """

    def __init__(self) -> None:
        """Initialise the stream manager."""
        self._streams: t.Dict[str, StreamIterator] = {}
        self._lock: asyncio.Lock = asyncio.Lock()

    async def create_stream(
        self,
        request_id: str,
        model: str,
        *,
        timeout: float = 60.0,
    ) -> StreamIterator:
        """Create a new stream iterator and register it.

        Args:
            request_id: The UUID of the streaming request.
            model: The model identifier.
            timeout: Timeout for each chunk read.

        Returns:
            A new StreamIterator instance.

        Raises:
            ValueError: If a stream with the same request_id already exists.
        """
        async with self._lock:
            if request_id in self._streams:
                raise ValueError(
                    f"Stream '{request_id}' already exists"
                )

            queue: asyncio.Queue[StreamEvent] = asyncio.Queue(
                maxsize=_MAX_QUEUE_SIZE
            )
            iterator: StreamIterator = StreamIterator(
                request_id,
                model,
                queue,
                timeout=timeout,
            )
            self._streams[request_id] = iterator
            return iterator

    async def get_stream(self, request_id: str) -> t.Optional[StreamIterator]:
        """Get a stream iterator by request ID.

        Args:
            request_id: The UUID of the streaming request.

        Returns:
            The StreamIterator, or None if not found.
        """
        async with self._lock:
            return self._streams.get(request_id)

    async def remove_stream(self, request_id: str) -> None:
        """Remove a stream from the manager.

        Args:
            request_id: The UUID of the stream to remove.
        """
        async with self._lock:
            self._streams.pop(request_id, None)

    def list_active_streams(self) -> t.List[t.Dict[str, t.Any]]:
        """List all active streams with their stats.

        Returns:
            A list of stream statistics dictionaries.
        """
        return [
            stream.get_stats()
            for stream in self._streams.values()
            if not stream.stream_finished
        ]

    async def cancel_all(self) -> None:
        """Cancel all active streams."""
        async with self._lock:
            for stream in self._streams.values():
                if not stream.stream_finished:
                    await stream.cancel()
            self._streams.clear()

    @property
    def active_count(self) -> int:
        """Number of active (non-finished) streams."""
        return sum(
            1 for s in self._streams.values() if not s.stream_finished
        )

    @property
    def total_count(self) -> int:
        """Total number of registered streams (including finished)."""
        return len(self._streams)

    def __repr__(self) -> str:
        """Return a string representation of the stream manager."""
        return (
            f"StreamManager(active={self.active_count}, "
            f"total={self.total_count})"
        )


# ---------------------------------------------------------------------------
# Stream event helpers
# ---------------------------------------------------------------------------


def create_chunk_event(chunk: InferenceChunk) -> StreamEvent:
    """Create a chunk event for the stream queue.

    Args:
        chunk: The inference chunk.

    Returns:
        A StreamEvent with type ``"chunk"``.
    """
    return StreamEvent(type="chunk", chunk=chunk)


def create_error_event(error: BaseException) -> StreamEvent:
    """Create an error event for the stream queue.

    Args:
        error: The exception that occurred.

    Returns:
        A StreamEvent with type ``"error"``.
    """
    return StreamEvent(type="error", error=error)


def create_done_event(final_data: t.Optional[JSONObject] = None) -> StreamEvent:
    """Create a done event for the stream queue.

    Args:
        final_data: Optional final response data.

    Returns:
        A StreamEvent with type ``"done"``.
    """
    return StreamEvent(type="done", final_data=final_data)


# ---------------------------------------------------------------------------
# Stream chunk parsing
# ---------------------------------------------------------------------------


def parse_stream_chunk(data: JSONObject) -> InferenceChunk:
    """Parse a stream message data payload into an InferenceChunk.

    The daemon sends stream chunks with the structure::

        {
            "type": "stream",
            "id": "<uuid>",
            "data": {
                "token": "...",
                "final": false,
                ...
            }
        }

    Args:
        data: The ``data`` field of a stream message.

    Returns:
        A parsed InferenceChunk.

    Raises:
        ValueError: If the data is malformed.
    """
    return InferenceChunk.from_dict(data)


# ---------------------------------------------------------------------------
# Accumulator for collecting stream results
# ---------------------------------------------------------------------------


class StreamAccumulator:
    """Accumulates streaming chunks into a complete response.

    This is useful when you want to collect the full output while also
    processing tokens as they arrive.

    Usage::

        accumulator = StreamAccumulator()
        async for chunk in client.infer_stream("model", "Hello"):
            accumulator.add(chunk)
            print(chunk.token, end="", flush=True)
        print(f"\\nFull text: {accumulator.text}")
    """

    def __init__(self) -> None:
        """Initialise the accumulator."""
        self._tokens: t.List[str] = []
        self._chunks: t.List[InferenceChunk] = []
        self._final_chunk: t.Optional[InferenceChunk] = None
        self._start_time: float = timestamp()

    def add(self, chunk: InferenceChunk) -> None:
        """Add a chunk to the accumulator.

        Args:
            chunk: The inference chunk to add.
        """
        self._chunks.append(chunk)
        if chunk.token:
            self._tokens.append(chunk.token)
        if chunk.final:
            self._final_chunk = chunk

    @property
    def text(self) -> str:
        """The accumulated text (all tokens concatenated)."""
        return "".join(self._tokens)

    @property
    def token_count(self) -> int:
        """Number of tokens accumulated."""
        return len(self._tokens)

    @property
    def finish_reason(self) -> t.Optional[str]:
        """The finish reason (if the stream has completed)."""
        if self._final_chunk is not None:
            return self._final_chunk.finish_reason
        return None

    @property
    def usage(self) -> t.Optional[TokenUsage]:
        """Token usage statistics (if available)."""
        if self._final_chunk is not None:
            return self._final_chunk.usage
        return None

    @property
    def elapsed(self) -> float:
        """Elapsed time since the accumulator was created."""
        return timestamp() - self._start_time

    def reset(self) -> None:
        """Reset the accumulator to its initial state."""
        self._tokens.clear()
        self._chunks.clear()
        self._final_chunk = None
        self._start_time = timestamp()

    def __repr__(self) -> str:
        """Return a string representation of the accumulator."""
        return (
            f"StreamAccumulator(tokens={self.token_count}, "
            f"finished={self._final_chunk is not None})"
        )


__all__: list[str] = [
    "StreamIterator",
    "StreamManager",
    "StreamAccumulator",
    "StreamEvent",
    "create_chunk_event",
    "create_error_event",
    "create_done_event",
    "parse_stream_chunk",
]

File: bindings/node/src/stream.ts
/**
 * Ainos SDK — Streaming support for inference responses.
 *
 * Provides a {@link InferenceStream} class that wraps the raw TCP transport
 * into a convenient EventEmitter interface.  Streams emit `data` events for
 * each text chunk, a `progress` event with cumulative stats, and `end` /
 * `error` terminal events.
 */

import { EventEmitter } from 'events';
import { InferenceChunk } from './types';
import { TcpTransport } from './transport';
import { safeDecodeJson } from './utils';
import { InferenceError } from './errors';

// ============================================================================
// Constants
// ============================================================================

/** Default high-water mark for backpressure (number of buffered chunks). */
const DEFAULT_HIGH_WATER_MARK = 1024;

// ============================================================================
// InferenceStream
// ============================================================================

/**
 * A streaming inference session.
 *
 * Emits:
 * - `data` (chunk: string) — for each text fragment received
 * - `progress` (tokensGenerated: number, elapsedMs: number) — cumulative stats
 * - `end` () — when the stream completes
 * - `error` (err: Error) — when an error occurs
 *
 * Usage:
 * ```ts
 * const stream = client.inferStream({ prompt: "Hello" });
 * stream.on('data', (chunk) => process.stdout.write(chunk));
 * stream.on('end', () => console.log('\n[Done]'));
 * ```
 */
export class InferenceStream extends EventEmitter {
  private transport: TcpTransport;
  private request: Record<string, unknown>;
  private _started = false;
  private _ended = false;
  private _tokensGenerated = 0;
  private _startTime = 0;
  private buffer: string[] = [];
  private highWaterMark: number;

  // Backpressure tracking
  private _paused = false;
  private drainResolve: (() => void) | null = null;

  constructor(
    transport: TcpTransport,
    request: Record<string, unknown>,
    options?: { highWaterMark?: number },
  ) {
    super();
    this.transport = transport;
    this.request = request;
    this.highWaterMark = options?.highWaterMark ?? DEFAULT_HIGH_WATER_MARK;
  }

  // --------------------------------------------------------------------------
  // Properties
  // --------------------------------------------------------------------------

  /** Whether the stream has started receiving data. */
  get started(): boolean {
    return this._started;
  }

  /** Whether the stream has ended. */
  get ended(): boolean {
    return this._ended;
  }

  /** Total tokens generated so far (from the daemon's reporting). */
  get tokensGenerated(): number {
    return this._tokensGenerated;
  }

  /** Elapsed milliseconds since the stream started. */
  get elapsedMs(): number {
    if (!this._startTime) return 0;
    return Date.now() - this._startTime;
  }

  // --------------------------------------------------------------------------
  // Public API
  // --------------------------------------------------------------------------

  /**
   * Start the streaming inference.
   *
   * This sends the inference request and begins reading the response stream.
   * The stream must be consumed via event listeners.
   */
  start(): void {
    if (this._started) {
      throw new Error('Stream already started');
    }
    this._started = true;
    this._startTime = Date.now();

    // Send the request
    const payload = { ...this.request, type: 'InferenceStream' };
    this.transport.send(payload);

    // Begin reading lines in the background
    this.readNextChunk();
  }

  /**
   * Cancel the stream before it completes.
   */
  cancel(): void {
    if (this._ended) return;
    this._ended = true;
    this.buffer = [];
    this.emit('end');
    this.removeAllListeners();
  }

  /**
   * Pause the stream (backpressure).
   *
   * While paused, chunks are buffered internally.  Call {@link resume} to
   * resume emitting `data` events.
   */
  pause(): void {
    this._paused = true;
  }

  /**
   * Resume a paused stream.
   */
  resume(): void {
    if (!this._paused) return;
    this._paused = false;

    // Flush buffered chunks
    while (this.buffer.length > 0 && !this._paused && !this._ended) {
      const chunk = this.buffer.shift()!;
      this.emit('data', chunk);
    }

    // Signal drain
    if (this.drainResolve) {
      this.drainResolve();
      this.drainResolve = null;
    }

    // If we drained the buffer, resume reading
    if (this.buffer.length === 0 && !this._ended) {
      this.readNextChunk();
    }
  }

  /**
   * Return a promise that resolves when the stream ends.
   */
  waitForEnd(): Promise<void> {
    return new Promise((resolve, reject) => {
      if (this._ended) {
        resolve();
        return;
      }
      this.once('end', resolve);
      this.once('error', reject);
    });
  }

  // --------------------------------------------------------------------------
  // Internal: Chunk Reading
  // --------------------------------------------------------------------------

  /**
   * Read the next available chunk line from the transport.
   *
   * This is called recursively until the stream ends.
   */
  private async readNextChunk(): Promise<void> {
    if (this._ended) return;

    try {
      // Apply backpressure: if paused and buffer is full, wait
      if (this._paused && this.buffer.length >= this.highWaterMark) {
        await new Promise<void>((resolve) => {
          this.drainResolve = resolve;
        });
        if (this._ended) return;
      }

      const line = await this.transport.sendAndReceive(this.request, 300000);

      if (this._ended) return;

      const parsed = safeDecodeJson<{ type: string }>(line, { type: '' });

      if (!parsed || !parsed.type) {
        // Malformed line — skip
        this.readNextChunk();
        return;
      }

      switch (parsed.type) {
        case 'InferenceChunk': {
          const chunk = safeDecodeJson<InferenceChunk>(line, { chunk: '', done: false });
          if (chunk.chunk.length > 0) {
            this.emitChunk(chunk.chunk);
          }
          if (chunk.done) {
            this.finish();
          } else {
            this.readNextChunk();
          }
          break;
        }

        case 'Error': {
          const errData = safeDecodeJson<{ message: string }>(line, { message: 'Unknown streaming error' });
          this.fail(new InferenceError(errData.message));
          break;
        }

        case 'InferenceResponse': {
          // Non-streaming response returned for a stream request — treat as a single chunk
          const respData = safeDecodeJson<{ output: string; tokens_generated?: number }>(
            line,
            { output: '', tokens_generated: 0 },
          );
          if (respData.output.length > 0) {
            this.emitChunk(respData.output);
          }
          this._tokensGenerated = respData.tokens_generated ?? 0;
          this.finish();
          break;
        }

        default: {
          // Unknown type — skip and continue
          this.readNextChunk();
          break;
        }
      }
    } catch (err) {
      if (!this._ended) {
        this.fail(err instanceof Error ? err : new Error(String(err)));
      }
    }
  }

  /**
   * Emit a chunk, applying backpressure.
   */
  private emitChunk(chunk: string): void {
    if (this._paused) {
      this.buffer.push(chunk);
      return;
    }

    this.emit('data', chunk);

    // Update progress
    this._tokensGenerated += 1;
    this.emit('progress', this._tokensGenerated, this.elapsedMs);

    // Check if we need to pause
    if (this.listenerCount('data') === 0) {
      // No consumer — buffer
      this.buffer.push(chunk);
      if (this.buffer.length >= this.highWaterMark) {
        this._paused = true;
      }
    }
  }

  /**
   * Finish the stream successfully.
   */
  private finish(): void {
    if (this._ended) return;
    this._ended = true;
    this.emit('progress', this._tokensGenerated, this.elapsedMs);
    this.emit('end');
    this.removeAllListeners();
  }

  /**
   * Fail the stream with an error.
   */
  private fail(err: Error): void {
    if (this._ended) return;
    this._ended = true;
    this.emit('error', err);
    this.removeAllListeners();
  }
}

// ============================================================================
// ReadableStreamAdapter
// ============================================================================

/**
 * Adapt an {@link InferenceStream} to a Node.js `Readable` stream.
 *
 * This allows the streaming inference to be used with Node.js stream
 * pipelines and Web Streams API consumers.
 */
export class ReadableStreamAdapter extends (require('stream').Readable) {
  private stream: InferenceStream;

  constructor(stream: InferenceStream, options?: { highWaterMark?: number }) {
    super({
      objectMode: true,
      highWaterMark: options?.highWaterMark ?? DEFAULT_HIGH_WATER_MARK,
    });
    this.stream = stream;

    this.stream.on('data', (chunk: string) => {
      if (!this.push(chunk)) {
        this.stream.pause();
      }
    });

    this.stream.on('end', () => {
      this.push(null);
    });

    this.stream.on('error', (err: Error) => {
      this.destroy(err);
    });
  }

  _read(): void {
    this.stream.resume();
  }
}

// ============================================================================
// Stream Accumulator
// ============================================================================

/**
 * Accumulate a streaming inference into a single string.
 *
 * Useful for testing or when you want the full output after streaming.
 *
 * @param stream - The inference stream to consume.
 * @returns A promise that resolves to the full concatenated output.
 */
export function accumulateStream(stream: InferenceStream): Promise<string> {
  return new Promise((resolve, reject) => {
    const parts: string[] = [];
    stream.on('data', (chunk: string) => parts.push(chunk));
    stream.on('end', () => resolve(parts.join('')));
    stream.on('error', reject);
    stream.start();
  });
}